common/cache.go

```diff
@@ -13,90 +18,300 @@ import (
 
 type blockCacheEntry struct {
 	data []byte
-	hash []byte
 }
 
-// BlockCache contains a set of recent compact blocks in marshalled form.
+// BlockCache contains a consecutive set of recent compact blocks in marshalled form.
 type BlockCache struct {
-	MaxEntries int
+	lengthsName, blocksName string // pathnames
+	lengthsFile, blocksFile *os.File
+	starts                  []int64 // Starting offset of each block within blocksFile
+	firstBlock              int     // height of the first block in the cache (usually Sapling activation)
+	nextBlock               int     // height of the first block not in the cache
+	latestHash              []byte  // hash of the most recent (highest height) block, for detecting reorgs.
+	mutex                   sync.RWMutex
+}
+
+func (c *BlockCache) GetNextHeight() int {
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+	return c.nextBlock
+}
+
+func (c *BlockCache) GetLatestHash() []byte {
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+	return c.latestHash
+}
+
+//  HashMismatch indicates if the given prev-hash doesn't match the most recent block's hash
+// so reorgs can be detected.
+func (c *BlockCache) HashMismatch(prevhash []byte) bool {
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+	return c.latestHash != nil && !bytes.Equal(c.latestHash, prevhash)
+}
+
+// Start over, so that all blocks get re-downloaded (file corruption)
+func (c *BlockCache) truncZero() {
```

Comment:
Requested change: This should always generate a log message, it looks like there's a log message in every case of detecting corruption, but it would be nice to have a single consistent log message that sysadmins can set up their alerting to detect. It would also be great to document the log message in the README, so everyone running this in a production environment is aware this can happen and what to look for in the logs.